### PR TITLE
[proxy] weavewait rewrites /etc/hosts so weave ip is returned first from 'hostname -i'

### PR DESCRIPTION
**Pull request by paulbellamy**
Partial fix for #68. Won't help with dynamic `attach/detach` or the weave script.

### CODE REVIEW

**Comment by rade** [comment]
I think this should close #68 and we should open a new issue (or issues) covering the remaining cases.

@bboreham would you mind reviewing this?

**Comment by paulbellamy** [comment]
I've tried this with Java's `InetAddress.getLocalHost()`, and it returns the weave address.

```bash
$ weave launch
$ docker $(weave proxy-config) run -ti --rm my-java-app
2c80b6ca84b7/10.128.0.2
$ docker run -ti --rm my-java-app
fc551c2b55a7/172.17.0.5
```

File: prog/weavewait/main.go [hunk 1]
```diff
@@ -39,3 +45,33 @@ func checkErr(err error) {
...
+	addrs, err := iface.Addrs()
+	checkErr(err)
+	if len(addrs) == 0 {
+		return
+	}
+	hostname, err := os.Hostname()
+	checkErr(err)
+	var oldHosts, newHosts bytes.Buffer
+	for _, addr := range addrs {
+		fmt.Fprint(&newHosts, strings.Split(addr.String(), "/")[0], "\t", hostname, "\n")
```
Comment (bboreham) [inline prog/weavewait/main.go]: Might be simpler to type-check addr as an `IPNet` and fetch the address directly; if we encounter some other type of address at this point we don't really want to go string-manipulating it.

File: test/610_proxy_wait_for_weave_test.sh [hunk 2]
```diff
@@ -19,4 +19,7 @@ COMMITTED_IMAGE=$(proxy docker_on $HOST1 commit c1)
 assert_raises "proxy docker_on $HOST1 run --name c2 $COMMITTED_IMAGE"
 assert "entrypoint c2" "$(entrypoint $COMMITTED_IMAGE)"
 
+# Check containers get their weave IP prepended to /etc/hosts
```
Comment (bboreham) [inline test/610_proxy_wait_for_weave_test.sh]: Maybe better to comment the effect rather than the mechanism (which isn't apparent here nor important)

File: prog/weavewait/main.go (hunk 1 above)
Comment (bboreham) [inline prog/weavewait/main.go]: Also I would question the value of putting the ethwe ipv6 address into `/etc/hosts`

File: test/610_proxy_wait_for_weave_test.sh [hunk 3]
```diff
@@ -19,4 +19,7 @@ COMMITTED_IMAGE=$(proxy docker_on $HOST1 commit c1)
 assert_raises "proxy docker_on $HOST1 run --name c2 $COMMITTED_IMAGE"
 assert "entrypoint c2" "$(entrypoint $COMMITTED_IMAGE)"
 
+# Check weave IP is first ip returned, so java (etc) prefer weave.
+assert "proxy docker_on $HOST1 run -e 'WEAVE_CIDR=10.2.1.1/24' $BASE_IMAGE hostname -i | cut -d' ' -f1" "10.2.1.1"
```
Comment (bboreham) [inline test/610_proxy_wait_for_weave_test.sh]: In the presence of multiple addresses, `hostname -i` does not seem to guarantee any particular order.  By changing the first entry (added by `weavewait`) to 200.128.0.2 I can make it come second.

**Comment by paulbellamy** [comment]
As @bboreham points out above, afaict `hostname -i` makes no guarantees about order. Nor does Java.

If we want to add the `ip hostname` into `/etc/hosts` we could use docker's `--add-hosts`, but as `hostname -I` already returns the weave IP, I don't see the point of that.

**Comment by rade** [comment]
`--add-host` was considered in #68. It is of no use. The whole point of #68 is make `hostname -i` return the weave address. first.

We could *remove* the host entry for the docker ip.

**Comment by paulbellamy** [comment]
Yeah. removing the docker ip seems a bit naughty..

**Comment by rade** [comment]
> Yeah. removing the docker ip seems a bit naughty.

Meh. The docker ip should still show up with `hostname -I`. Provided that both inbound and outbound docker links still work that is fine, I reckon. And we can add a flag for disabling that behaviour.

**Comment by paulbellamy** [comment]
Tried out links with the docker IP removed from /etc/hosts, and they still work fine. Not a big fan of adding yet another flag, but...

File: prog/weavewait/main.go [hunk 4]
```diff
@@ -1,32 +1,54 @@
...
 	}
-	args := os.Args[1:]
+
+	var (
+		args         = os.Args[1:]
+		notInExec    = true
+		rewriteHosts = true
+	)
 
 	if args[0] == "-s" {
```
Comment (tomwilkie) [inline prog/weavewait/main.go]: Seems like this is getting to the stage where it might need 'proper' argument parsing, usage etc.

File: prog/weavewait/main.go [hunk 5]
```diff
@@ -39,3 +61,43 @@ func checkErr(err error) {
...
+	checkErr(err)
+	if strings.Contains(oldHosts.String(), newHosts.String()) {
+		return
+	}
+	_, err = f.Seek(0, 0)
+	checkErr(err)
+	written, err := io.Copy(f, &newHosts)
+	checkErr(err)
+	// discard the first line, if it is the docker host
+	firstLine, err := oldHosts.ReadString('\n')
```
Comment (tomwilkie) [inline prog/weavewait/main.go]: This seems to discard the first line if it isn't 127.0.0.1?

Comment (paulbellamy) [inline prog/weavewait/main.go]: correct. We remove the docker ip from /etc/hosts, so that it doesn't show up in `hostname -i`.

File: prog/weavewait/main.go [hunk 6]
```diff
@@ -39,3 +61,43 @@ func checkErr(err error) {
 		os.Exit(1)
 	}
 }
+
+// Prepend the addresses to /etc/hosts, removing the docker address
+func updateHosts(iface *net.Interface) {
```
Comment (tomwilkie) [inline prog/weavewait/main.go]: I think it might be best to try a different approach with this function.  This current version is too hard to follow and maintain IMO.

I would construct, in memory, what you want the file to be, maybe as a [][2]string.  Read the current contents, remove the entries you don't want, add the entries you do, sort(?) then write out to disk.

File: prog/weavewait/main.go (hunk 4 above)
Comment (paulbellamy) [inline prog/weavewait/main.go]: Yeah. Pushed that off until next time it changes.

File: prog/weavewait/main.go [hunk 7]
```diff
@@ -39,3 +63,79 @@ func checkErr(err error) {
...
+		if addr, ok := addr.(*net.IPNet); ok {
+			ip := addr.IP.String()
+			hosts[ip] = append(hosts[ip], hostname)
+		}
+	}
+
+	writeHosts(hosts)
+}
+
+func parseHosts() map[string][]string {
```
Comment (tomwilkie) [inline prog/weavewait/main.go]: This returns a map of ip -> []hostname, right?

I wonder if you made this hostname -> []ip, the code above might be simpler?

Comment (paulbellamy) [inline prog/weavewait/main.go]: The code above would be simpler, but writeHosts would be more complicated.

Comment (tomwilkie) [inline prog/weavewait/main.go]: Do you have to output the ips in sorted order?

**Comment by tomwilkie** [comment]
LGTM

File: prog/weavewait/main.go [hunk 8]
```diff
@@ -39,3 +63,79 @@ func checkErr(err error) {
...
+		line := scanner.Text()
+
+		// Remove any comments
+		if i := strings.IndexByte(line, '#'); i != -1 {
+			line = line[:i]
+		}
+
+		fields := strings.Fields(line)
+		if len(fields) > 0 {
+			ips[fields[0]] = fields[1:]
```
Comment (bboreham) [inline prog/weavewait/main.go]: I think it would be safer to append here, in case someone has a hosts file with the same IP address twice.

File: prog/weavewait/main.go [hunk 9]
```diff
@@ -39,3 +63,79 @@ func checkErr(err error) {
 		os.Exit(1)
 	}
 }
+
+// Prepend the addresses to /etc/hosts, removing the docker address
```
Comment (bboreham) [inline prog/weavewait/main.go]: I don't think it does prepend.

File: prog/weavewait/main.go [hunk 10]
```diff
@@ -39,3 +63,79 @@ func checkErr(err error) {
...
+	}
+	sort.Strings(ips)
+
+	buf := &bytes.Buffer{}
+	for _, ip := range ips {
+		if addrs := contents[ip]; len(addrs) > 0 {
+			fmt.Fprintf(buf, "%s\t%s\n", ip, strings.Join(addrs, " "))
+		}
+	}
+	checkErr(ioutil.WriteFile("/etc/hosts", buf.Bytes(), 644))
```
Comment (bboreham) [inline prog/weavewait/main.go]: Since you have removed all pre-existing comments and re-ordered the lines, it might be a good idea to add one new comment like "# modified by weave", for the benefit of puzzled users.

File: test/610_proxy_wait_for_weave_test.sh (hunk 3 above)
Comment (bboreham) [inline test/610_proxy_wait_for_weave_test.sh]: Now we are removing the weave IP the test is safe, but the comment in the test still refers to checking the "first ip".  I think it should be the only IP.
